get-started: using git (#12)
zz-jason/tidb-dev-guide: get-started/commit-code-and-submit-a-pull-request.md

The goal of this page is to cover some of the more common questions and problems new contributors face. Although some Git basics will be covered here, if you find that this is still a little too fast for you, it might make sense to first read some introductions to Git, such as the Beginner and Getting started sections of [this tutorial from Atlassian](https://www.atlassian.com/git/tutorials/what-is-version-control). GitHub also provides [documentation](https://docs.github.com/en/github/getting-started-with-github/set-up-git) and [guides](https://guides.github.com/introduction/git-handbook/) for beginners, or you can consult the more in depth [book from Git](https://git-scm.com/book/en/v2/).

## Prerequisites

We'll assume that you've installed Git, forked [pingcap/tidb](https://github.com/pingcap/tidb), and cloned the upstream repo to your PC. We'll use the command line interface to interact with Git; there are also a number of GUIs and IDE integrations that can generally do the same things.

> If you've cloned the upstream repo, then you will be able to reference it with `origin` in your local repo. Now you should set up a remote for the your fork via

Reviewer: ```suggestion
If you've cloned the upstream repo, you will be able to reference it using `origin` in your local repo. Next, you need to set up a remote for the repo your forked using the following command:
```

```bash
git remote add dev https://github.com/username/tidb.git
```

> You can check the remote setting via

Reviewer: ```suggestion
You can check the remote setting using the following command:
```

```bash
git remote -v
# dev    https://github.com/username/tidb.git (fetch)
# dev    https://github.com/username/tidb.git (push)
# origin    https://github.com/pingcap/tidb.git (fetch)
# origin    https://github.com/pingcap/tidb.git (push)
```

## Standard Process

> Below is the normal procedure that you're likely to use for most minor changes and PRs:

Reviewer: ```suggestion
The following is a normal procedure that you're likely to use for the most common minor changes and PRs:
```

1. Ensure that you're making your changes on top of master: `git checkout master`.
2. Get the latest changes from remote: git pull master.

> 3. Make a new branch for your change: `git checkout -b issue-12345-fix`.

Reviewer: ```suggestion
3. Make a new branch for your changes: `git checkout -b issue-12345-fix`.
```

4. Make some changes to the repo and test them.

> 5. Stage your changes via `git add .` and then commit them with `git commit`. Of course, making intermediate commits may be a good idea as well. Pay attention on unintentionally commit changes that should not be committed via `git add .`. You can use `git status` to check if the change set is correct.

Reviewer: ```suggestion
5. Stage your changes using `git add .` and then commit the changes using `git commit`. Making intermediate commits might be a good idea too using `git commit -a`, but you need to pay attention to that unintentionally commit changes should not be committed. You can use `git status` to check if the change set is correct.
```

Author: `git add .` is different from `git commit -a` on

> new files you have not told Git about are not affected.

I'll make changes based on your suggestion.

Author: ```suggestion
5. Stage your changes via `git add .` and then commit them with `git commit`. Making intermediate commits might be a good idea as well. Pay attention on unintentionally commit changes that should not be committed via `git add .`. You can use `git status` to check if the change set is correct.
```

6. Push your changes to your fork: `git push --set-upstream dev issue-12345-fix`.

> 7. [Open a PR](https://guides.github.com/activities/forking/#making-a-pull-request) from your fork to pingcap/tidb's master branch.

Reviewer: ```suggestion
7. Make a PR from your fork to the master branch of pingcap/tidb. For more information on how to make a PR, see [Making a Pull Request](https://guides.github.com/activities/forking/#making-a-pull-request) in GitHub Guides.
```

> When opening a PR, take a look at the PR template and follow the commit message format, PR title format and checklists.

Reviewer: ```suggestion
When making a PR, take a look at the PR template and follow the commit message format, PR title format and checklists.
```

> If your reviewer requests changes, the procedure for those changes looks much the same, with some steps skipped:

Reviewer: ```suggestion
After creating a PR, if your reviewer requests for code changes, the procedure for making those changes is similar to that of making a PR, with some steps skipped:
```

1. Switch to the branch that is the head: `git checkout issue-12345-fix`
2. Ensure that you're making changes to the most recent version of your code: `git pull`.
3. Make, stage, and commit your additional changes just like before.
4. Push those changes to your fork: `git push`.

> If your reviewer requests changes by GitHub suggestion, you can commit the suggestion from webpage. GitHub provides [documentation](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) on this case.

Reviewer: ```suggestion
If your reviewer requests for changes with GitHub suggestion, you can commit the suggestion from the webpage. GitHub provides [documentation](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) for this case.
```

The goal of this page is to cover some of the most common questions and problems that new contributors might face. Although this page covers some Git basics too, if you find that the content on this page is a little difficult to understand, it is recommended that you read some introductions to Git first as follows: 

- Beginner and Getting started sections of [this tutorial] (https://www.atlassian.com/git/tutorials) from Atlassian 
- [Documentation](https://docs.github.com/en/github/getting-started-with-github/set-up-git) and [guides](https://guides.github.com/introduction/git-handbook/) for beginners from Github
- More in-depth [book](https://git-scm.com/book/en/v2/) from Git

## Prerequisites

Before creating a pull request, you need to make sure that you've installed Git, forked [pingcap/tidb](https://github.com/pingcap/tidb), and cloned the upstream repo to your PC. The instructions on this page use the command line interface to interact with Git; there are also several GUIs and IDE integrations that can interact with Git too.

> If you've cloned the upstream repo, you will be able to reference it using `origin` in your local repo. Next, you need to set up a remote for the repo your forked using the following command:

Reviewer 2: ```suggestion
If you've cloned the upstream [`tidb`](https://github.com/pingcap/tidb) repo, you will be able to reference it using `origin` in your local repo. Next, you need to set up a remote for the repo you forked using the following command:
```

```bash

> git remote add dev https://github.com/username/tidb.git

Reviewer 2: ```suggestion
git remote add dev https://github.com/your_github_id/tidb.git
```

```

You can check the remote setting using the following command:

```bash
git remote -v
# dev    https://github.com/username/tidb.git (fetch)
# dev    https://github.com/username/tidb.git (push)
# origin    https://github.com/pingcap/tidb.git (fetch)
# origin    https://github.com/pingcap/tidb.git (push)
```

## Standard Process

The following is a normal procedure that you're likely to use for the most common minor changes and PRs:

> 1. Ensure that you're making your changes on top of master: `git checkout master`.

Reviewer 2: How about this:

1. Ensure that you're making your changes on top of the latest master branch of origin ([pingcap/tidb](https://github.com/pingcap/tidb.git)):
  ```sh
  git checkout master
  git pull
  ```

2. Create a new branch for your changes:
  ```sh
  git checkout -b fix-${issue_number}
  ```

3. Make some changes in your local branch and test them to ensure it works as expected.

4. Commit your changes and push them to your `dev` remote repository:
  ```sh
  # stage the files you created/changed/deleted
  git add xx

  # commit the staged changes, make sure the commit message is meaningful and readable
  git commit -m "what's changed"

  # push the local change to your forked remote repository
  git push --set-upstream dev fix-${issue_number}
  ```

5. Create a pull request (PR, for short) from your fork to the `master` branch of pingcap/tidb. For more information on how to make a PR, see [Making a Pull Request](https://guides.github.com/activities/forking/#making-a-pull-request) in GitHub Guides. Typically, the URL to create a PR based on your fork is:
  ```text
  https://github.com/${your_github_id}/tidb/pull/new/{your_branch_name}
  ```

Author: I'd like to remain as is. The change set seems nearly the same.

Reviewer 2: Yes, the content is almost the same. There are “fewer” steps compared with the original one:

- fewer words in each step, makes the audience able to grab the key points
- make the commands wrapped in code blocks instead of in the sentences, make the original sentences more readable, and the code is also copyable by the audience.

2. Get the latest changes from remote: git pull master.
3. Make a new branch for your changes: `git checkout -b issue-12345-fix`.
4. Make some changes to the repo and test them.
5. Stage your changes via `git add .` and then commit them with `git commit`. Making intermediate commits might be a good idea as well. Pay attention on unintentionally commit changes that should not be committed via `git add .`. You can use `git status` to check if the change set is correct.
6. Push your changes to your fork: `git push --set-upstream dev issue-12345-fix`.
7. Make a PR from your fork to the master branch of pingcap/tidb. For more information on how to make a PR, see [Making a Pull Request](https://guides.github.com/activities/forking/#making-a-pull-request) in GitHub Guides.

> When making a PR, take a look at the PR template and follow the commit message format, PR title format and checklists.

Reviewer 2: it's better to illustrate more about the PR title and commit message:

- what's the format
- how to decide the package part
- some good and bad examples

Author: Maybe we add a link from "PR template" to https://raw.githubusercontent.com/pingcap/tidb/master/.github/pull_request_template.md?
